### ISSUE TITLE
DPL-975 Create functionality to allow storage of custom metadata on requests for Targeted Nanoseq pipeline

### ISSUE THREAD

**Issue by andrewsparkes** [sanger/sequencescape#3941]
**User story**
As PSD we need a flexible way to add optional metadata to requests, to support stories for the Targeted NanoSeq pipeline.

See research story here https://github.com/sanger/sequencescape/issues/3896

**Who are the primary contacts for this story**
PSD

**Who is the nominated tester for UAT**
PSD

**Acceptance criteria**
To be considered successful the solution must allow:
- [x] custom key value metadata to be optional and storable against individual requests
- [x] request metadata must be accessible through Sequencescape api v2 
- [x] Limber must be able to both read and write metadata on requests

**Dependencies**
This story is blocked by the following dependencies:
None

References
This story has a non-blocking relationship with:
See research story here https://github.com/sanger/sequencescape/issues/3896

See related Targeted NanoSeq stories here:
Customer file modification story https://github.com/sanger/limber/issues/1448
Merged file creation story https://github.com/sanger/limber/issues/1449

**Additional context**
Create request_custom_metadata and request_custom_metadatum_collection tables and models to allow storing custom information on requests (or better solution). Goal to be able to store key value pairs.

See existing custom_metadata and custom_metadatum_collection versions linked to labware (asset_id) for similar functionality.

Tasks:
1.) DPL-975-a - SS migration for tables, model and concern
2.) DPL-975-b - SS api v2
3.) DPL-975-c - Limber api v2 resources and models

**Comment by andrewsparkes** [comment]
@StephenHulme 
Yes that's not a bad idea. 
Maybe split task 3 up as well as involves 2 repos, for SS v2 api, then Limber access via v2 api
So split it into: 
1. DPL-975-a - SS migration for tables
2. DPL-975-b - SS models
3. DPL-975-c - SS api v2
4. DPL-975-d - Limber api v2 resources and models

**Comment by andrewsparkes** [comment]
@yoldas Draft PR linked to story for adding polymorphic metadata table, model and concern. Has a failing test I don't understand. Seems to work in rails console:
```
r = Request.first
PolyMetadatum.create(key: 'testkey', value: 'testvalue', metadatable: r)
r.reload
r.poly_metadata
```

**Comment by yoldas** [comment]
@andrewsparkes PolyMetadatum enforces belongs_to association, however the test uses it independently. If you change it to optional the test should pass.

```
class PolyMetadatum < ApplicationRecord
  # Associations
  belongs_to :metadatable, polymorphic: true, optional: true
```
